File: src/sub_converter/parsers/vmess.py
"""VMess (vmess://) parser."""

from __future__ import annotations

import base64
import json
from typing import Any

from sub_converter.utils import pad_base64


def parse_vmess(url: str) -> dict[str, Any]:
    payload = url[len("vmess://"):]
    decoded = base64.b64decode(pad_base64(payload)).decode("utf-8", errors="strict")
    data = json.loads(decoded)
    data["scheme"] = "vmess"
    data["raw"] = url
    if "ps" in data and "name" not in data:
        data["name"] = data["ps"]
    return data

File: worker/src/parsers/vmess.ts
/**
 * VMess parser
 */

import { ProxyNode } from '../utils/types';

export function parseVMess(entry: string): ProxyNode {
  if (!entry.startsWith('vmess://')) {
    throw new Error('Invalid vmess URL');
  }

  const payload = entry.replace('vmess://', '');

  try {
    const padded = payload + '='.repeat((4 - (payload.length % 4)) % 4);
    const decoded = atob(padded);
    const config = JSON.parse(decoded);

    return {
      scheme: 'vmess',
      server: config.add || config.server,
      add: config.add || config.server,
      port: String(config.port),
      id: config.id,
      aid: config.aid || 0,
      scy: config.scy || 'auto',
      tls: config.tls,
      net: config.net || 'tcp',
      path: config.path,
      host: config.host,
      sni: config.sni,
      name: config.ps,
    };
  } catch (e) {
    throw new Error(`Failed to parse vmess: ${e}`);
  }
}
